hoprd/rest-api/src/session.rs

```diff
@@ -81,70 +128,289 @@ pub(crate) struct SessionClientResponse {
+    params(
+            ("protocol" = String, Path, description = "IP transport protocol")
+    ),
+    responses(
+            (status = 200, description = "Opened session listeners for the given IP protocol.", body = Vec<SessionClientResponse>),
+            (status = 400, description = "Invalid IP protocol.", body = ApiError),
+            (status = 401, description = "Invalid authorization token.", body = ApiError),
+            (status = 422, description = "Unknown failure", body = ApiError)
+    ),
+    security(
+            ("api_token" = []),
+            ("bearer_token" = [])
+    ),
+    tag = "Session",
+)]
+pub(crate) async fn list_clients(
+    State(state): State<Arc<InternalState>>,
+    Path(protocol): Path<IpProtocol>,
+) -> Result<impl IntoResponse, impl IntoResponse> {
+    let response = state
+        .open_listeners
+        .read()
+        .await
+        .iter()
+        .filter(|(id, _)| id.0 == protocol)
+        .map(|(id, (target, _))| SessionClientResponse {
+            protocol,
+            target: target.clone(),
+            ip: id.1.ip().to_string(),
+            port: id.1.port(),
+        })
+        .collect::<Vec<_>>();
+
+    Ok::<_, (StatusCode, ApiErrorStatus)>((StatusCode::OK, Json(response)).into_response())
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, utoipa::ToSchema)]
+#[schema(example = json!({
+        "listeningIp": "127.0.0.1",
+        "port": 5542
+    }))]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct SessionCloseClientRequest {
+    pub listening_ip: String,
+    pub port: u16,
+}
+
+/// Closes an existing Session listener.
+/// The listener must've been previously created and bound for the given IP protocol.
+/// Once a listener is closed, no more socket connections can be made to it.
+#[utoipa::path(
+    delete,
+    path = const_format::formatcp!("{BASE_PATH}/session/{{protocol}}"),
+    params(
+            ("protocol" = String, Path, description = "IP transport protocol")
+    ),
+    request_body(
+            content = SessionCloseClientRequest,
+            description = "Closes the listener on the given bound IP address and port.",
+            content_type = "application/json"),
+    responses(
+            (status = 204, description = "Listener closed successfully"),
+            (status = 400, description = "Invalid IP protocol or port.", body = ApiError),
+            (status = 401, description = "Invalid authorization token.", body = ApiError),
+            (status = 404, description = "Listener not found.", body = ApiError),
+            (status = 422, description = "Unknown failure", body = ApiError)
+    ),
+    security(
+            ("api_token" = []),
+            ("bearer_token" = [])
+    ),
+    tag = "Session",
+)]
+pub(crate) async fn close_client(
+    State(state): State<Arc<InternalState>>,
+    Path(protocol): Path<IpProtocol>,
+    Json(SessionCloseClientRequest { listening_ip, port }): Json<SessionCloseClientRequest>,
```

Comment:
Why don't we identify the session not by the local port which was given to client on creation?

Reply:
Because you can have listeners at different interfaces with different ips. See above.